File: RideSync - Backend/src/main/java/com/example/carpool/service/RideService.java
package com.example.carpool.service;

import java.util.List;
import java.util.Optional;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.example.carpool.entity.Ride;
import com.example.carpool.entity.User;
import com.example.carpool.repository.RideRepository;

@Service
@Transactional
public class RideService {

    @Autowired
    private RideRepository rideRepository;

    // Create a new ride offer
    public Ride createRide(Ride ride) {
        return rideRepository.save(ride);
    }

    // Get a ride by its ID
    public Optional<Ride> getRideById(Long id) {
        return rideRepository.findById(id);
    }

    // Get all rides offered by a specific driver
    public List<Ride> getRidesByDriver(User driver) {
        return rideRepository.findByDriver(driver);
    }


    // Save a ride offer or update an existing one
    public Ride saveRide(Ride ride) {
        return rideRepository.save(ride);
    }
    
    public List<Ride> getAllRides() {
        return rideRepository.findAll();
    }
    
    public List<Ride> searchRides(String source, String destination) {
        return rideRepository.findByOriginIgnoreCaseAndDestinationIgnoreCase(source, destination);
    }


}

File: RideSync - Frontend/src/services/rideService.js
import axios from 'axios';
import { authHeader } from '../utils/authHeader';

const API_URL = 'http://localhost:8080/api/rides/';

// Offer a new ride
const offerRide = (rideData) => {
  return axios.post(API_URL + 'offer', rideData, { headers: authHeader() });
};

// Get all rides
const getAllRides = () => {
  return axios.get(API_URL + 'all', { headers: authHeader() });
};

// Get ride details by ID
const getRideById = (id) => {
  return axios.get(API_URL + id, { headers: authHeader() });
};

const searchRides = (source, destination) => {
  return axios.get(API_URL + `search?source=${source}&destination=${destination}`, {
    headers: authHeader()
  });
};

export default {
  offerRide,
  getAllRides,
  getRideById,
  searchRides,
}